beacon-chain/sync/subscriber_beacon_blocks.go

```diff
@@ -61,104 +59,29 @@ func (s *Service) beaconBlockSubscriber(ctx context.Context, msg proto.Message)
 	return err
 }
 
-// processSidecarsFromExecution retrieves (if available) sidecars data from the execution client,
+// processSidecarsFromExecutionFromBlock retrieves (if available) sidecars data from the execution client,
 // builds corresponding sidecars, save them to the storage, and broadcasts them over P2P if necessary.
-func (s *Service) processSidecarsFromExecution(ctx context.Context, block interfaces.ReadOnlySignedBeaconBlock) {
-	if block.Version() >= version.Fulu {
-		s.processDataColumnSidecarsFromExecution(ctx, block)
-		return
-	}
-
-	if block.Version() >= version.Deneb {
-		s.processBlobSidecarsFromExecution(ctx, block)
-		return
-	}
-}
-
-// processDataColumnSidecarsFromExecution retrieves (if available) data column sidecars data from the execution client,
-// builds corresponding sidecars, save them to the storage, and broadcasts them over P2P if necessary.
-func (s *Service) processDataColumnSidecarsFromExecution(ctx context.Context, roSignedBlock interfaces.ReadOnlySignedBeaconBlock) {
-	block := roSignedBlock.Block()
-
-	log := log.WithFields(logrus.Fields{
-		"slot":          block.Slot(),
-		"proposerIndex": block.ProposerIndex(),
-	})
-
-	kzgCommitments, err := block.Body().BlobKzgCommitments()
-	if err != nil {
-		log.WithError(err).Error("Failed to read commitments from block")
-		return
-	}
-
-	if len(kzgCommitments) == 0 {
-		// No blobs to reconstruct.
-		return
-	}
-
-	blockRoot, err := block.HashTreeRoot()
-	if err != nil {
-		log.WithError(err).Error("Failed to calculate block root")
-		return
-	}
-
-	log = log.WithField("blockRoot", fmt.Sprintf("%#x", blockRoot))
-
-	if s.cfg.dataColumnStorage == nil {
-		log.Warning("Data column storage is not enabled, skip saving data column, but continue to reconstruct and broadcast data column")
-	}
-
-	// When this function is called, it's from the time when the block is received, so in almost all situations we need to get the data column from EL instead of the blob storage.
-	sidecars, err := s.cfg.executionReconstructor.ReconstructDataColumnSidecars(ctx, roSignedBlock, blockRoot)
-	if err != nil {
-		log.WithError(err).Debug("Cannot reconstruct data column sidecars after receiving the block")
-		return
-	}
+func (s *Service) processSidecarsFromExecutionFromBlock(ctx context.Context, roSignedBlock interfaces.ReadOnlySignedBeaconBlock) {
+	if roSignedBlock.Version() >= version.Fulu {
+		key := fmt.Sprintf("%#x", roSignedBlock.Block().StateRoot())
+		if _, err, _ := s.columnSidecarsExecSingleFlight.Do(key, func() (interface{}, error) {
```

Comment:
what is the benefit of this new type singleflight.Group, the function executed looks a little unsafe but maybe because I don't understand this very much

Reply:
maybe i'm off i'm just seeing this function `func() (interface{}, error) {` will go look into how singleflight is used

Reply:
For information the function executed under singleflight returns`nil`. We don't use at all `interface{}`. Only `error`.

Reply:
oh wait is this used so that 
-When the block is received from gossip, or
-When a data column sidecar is received from gossip.
don't conflict?

Reply:
Yes exactly!
This function can be ran up to 129 times per block (when the block itself is received via gossip, and when up to 128 data column sidecars are received via gossip).

We need this function to run at most once. SingleFlight will ensure there this function does not run multiple times in parallel for the same root. (It would produce the exact same result and so it would only be a waste of CPU...)